indispensable-josef/src/commonMain/kotlin/at/asitplus/signum/indispensable/josef/JwsSigned.kt

```diff
@@ -27,57 +32,99 @@ data class JwsSigned(
     }
 
     override fun toString(): String {
         return "JwsSigned(header=$header" +
-                ", payload=${payload.encodeToString(Base64UrlStrict)}" +
+                ", payload=${payload}" +
                 ", signature=$signature" +
                 ", plainSignatureInput='$plainSignatureInput')"
     }
 
 
     companion object {
-        fun deserialize(it: String): KmmResult<JwsSigned> = catching {
-            val stringList = it.replace("[^A-Za-z0-9-_.]".toRegex(), "").split(".")
-            if (stringList.size != 3) throw IllegalArgumentException("not three parts in input: $it")
-            val headerInput = stringList[0].decodeToByteArray(Base64UrlStrict)
-            val header =
-                JwsHeader.deserialize(headerInput.decodeToString()).mapFailure { it.apply { printStackTrace() } }
-                    .getOrThrow()
-            val payload = stringList[1].decodeToByteArray(Base64UrlStrict)
-            val signature = stringList[2].decodeToByteArray(Base64UrlStrict)
-                .let { bytes ->
-                    when (val curve = header.algorithm.ecCurve) {
-                        null -> CryptoSignature.RSAorHMAC(bytes)
-                        else -> CryptoSignature.EC.fromRawBytes(curve, bytes)
-                    }
-                }
+        /**
+         * Deserializes the input, expected to contain a valid JWS (three Base64-URL strings joined by `.`),
+         * into a [JwsSigned] with `ByteArray` as the type of the payload.
+         */
+        inline fun deserialize(it: String): KmmResult<JwsSigned<ByteArray>> = catching {
+            val stringList = it.parseTo3Parts()
+            val inputParts = stringList.map { it.decodeToByteArray(Base64UrlStrict) }
+            val header = inputParts[0].toJwsHeader()
+            val payload = inputParts[1]
+            val signature = inputParts[2].toSignature(header.algorithm.ecCurve)
+            val plainSignatureInput = stringList[0] + "." + stringList[1]
+            JwsSigned(header, payload, signature, plainSignatureInput)
+        }
+
+        /**
+         * Deserializes the input, expected to contain a valid JWS (three Base64-URL strings joined by `.`),
+         * into a [JwsSigned] with [P] as the type of the payload.
+         */
+        inline fun <reified P : Any> deserialize(it: String, json: Json = Json): KmmResult<JwsSigned<P>> = catching {
+            val stringList = it.parseTo3Parts()
+            val inputParts = stringList.map { it.decodeToByteArray(Base64UrlStrict) }
+            val header = inputParts[0].toJwsHeader()
+            val payload: P = inputParts[1].run { json.decodeFromString<P>(this.decodeToString()) }
+            val signature = inputParts[2].toSignature(header.algorithm.ecCurve)
             val plainSignatureInput = stringList[0] + "." + stringList[1]
             JwsSigned(header, payload, signature, plainSignatureInput)
         }
 
+        fun String.parseTo3Parts(): List<String> {
```

Comment:
`private`?

Reply:
Can't be private, because the other method using this is inline and public

Reply:
still feels wrong to pollute `String`s extension function this way

Reply:
couldn't we just do
```kotlin
inline fun <reified P : Any> deserialize(it: String, json: Json = Json) =
    deserialize(it).mapCatching { JwsSigned(it.header, json.decodeFromString<P>(it.payload.decodeToString()), it.signature, it.plainSignatureInput) }
```